### ISSUE TITLE
[Connect 2.5] Workstreams - Design Review

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#3239]
We have to implement UI to review design inside works. And here we can discuss the requirements.

**Comment by maxceem** [comment]
1. In the new design, I see the UI for reviewing design submissions. But I don't see the UI for copilots to add such design submissions. How we would add them before letting a customer to review them?

**Comment by maxceem** [comment]
2. When we provide feedback we use a rich text editor as per design https://marvelapp.com/54cf844/screen/59214106
     That's good for copilots, so they can see the formatted text in Connect. Though as I understand copilots after share the feedback with the members. And I'm not sure how they do this. If they post the feedback on the forum then, formating would be lost. Thus it's better to not let it format here, as the inputted format would be different from what the memebrs would see. So if members would see just plain text, I would suggest to keep feedback as plain text here too. So when customer input the feedback they would format it accordingly.

**Comment by maxceem** [comment]
3. Customers can choose up to 5 works on this page https://marvelapp.com/54cf844/screen/59214106 or edit placement on this page https://marvelapp.com/54cf844/screen/59214115.
    How customers can mark the review as complete? I guess it shouldn't be automatically as we never know if a customer already did the final decision, or still thinking and wants to change anything.

**Comment by vikasrohit** [comment]
1. Good catch. For now, can we use our existing way to add design submission?

2. I think most of the copilots does not do simply copy paste of the feedback to the members. They parse the feedback and explain that to the members in probably a different way. So, I guess, we can keep the customer feedback formatted for the copilots/CAs.

3. Just fyi, https://marvelapp.com/54cf844/screen/59214106 is our current checkpoint-review (where we don't need the placement) and https://marvelapp.com/54cf844/screen/59214115 is our current final-designs (where placement matters).
Lets put a button for now to mark the review complete. May be we can put a button just below the General feedback and before the Individual feedback section. I am going to get more details from the product management team as well for this.

**Comment by maxceem** [comment]
> 1. Good catch. For now, can we use our existing way to add design submission?

I think that wouldn't work. In the old UI, we add one link per design. But in the new UI we should add 
- submission preview
- submission title
- several links

![image](https://user-images.githubusercontent.com/146016/62359352-35b7f880-b549-11e9-9456-17396578afc7.png)

Maybe I can run parallel challenges to upload the list of submission and to review submissions. If one challenge to do both things would be too big.
I can quickly create some design for such UI based on what we have before.

**Comment by vikasrohit** [comment]
Yep, that is my intent. I meant that we have to use something similar from our existing designs. I have asked the product managers and they are getting the designs worked by design team. So, we have to just assume a very simple design to achieve this for now, which we can improve later on, once we have the designs.

**Comment by maxceem** [comment]
@vikasrohit I'm proposing the next design for adding design links:

![image](https://user-images.githubusercontent.com/146016/62600404-5bfedf00-b921-11e9-88dc-d00e583f4690.png)

**Comment by vikasrohit** [comment]
I am good with the designs. Thanks for the extra efforts @maxceem

**Comment by maxceem** [comment]
5. Regarding the RichTextEditor which we should use during the review https://marvelapp.com/54cf844/screen/59214106. As I see, the component which we have in Connect App for editing RichText is quite hard to reuse and it's tailored for comments. So as it's not easy to reuse it, I would prefer not to include it to the challenge for implementing main review functionality as it's already quite complex. Would it be ok, to release 2.5 without rich text editor and add it later? Probably in 2.5.1 or 2.5.2.

**Comment by vikasrohit** [comment]
I think we would need the feature to manage the requirements in `2.5`. How difficult it is to simply use a new rich text editor which we evaluated earlier and just use that for Requirements tab of work card and general feedback of review milestone?

btw, I was looking at the `RichTextEditor` component and I found it not much coupled with comments except https://github.com/appirio-tech/connect-app/blob/feature/workstreams-workitems/src/components/RichTextArea/RichTextArea.jsx#L284 and https://github.com/appirio-tech/connect-app/blob/feature/workstreams-workitems/src/components/RichTextArea/RichTextArea.jsx#L141, which I think is easy to to refactor. Please correct me if I am missing something here.

**Comment by maxceem** [comment]
> I think we would need the feature to manage the requirements in 2.5. How difficult it is to simply use a new rich text editor which we evaluated earlier and just use that for Requirements tab of work card and general feedback of review milestone?

I'm not sure yet. So I would prefer to start with a single place on the requirements tab. And if works good, we may use it for other places.
Also, there is one tricky moment. Currently, our editor saves data in markdown format. The `ckeditor5` supports it as an experimental feature and still suggests to stick to the HTML format for storing richtext data https://ckeditor.com/docs/ckeditor5/latest/features/markdown.html.
So it's better to evaluate first how good it works with markdown format.

So I think I'll ask to reuse existent `RichTextEditor` for now for editing reviews. I think you are right, maybe it's not that hard to reuse it here.

**Comment by vikasrohit** [comment]
Good catch before implementing the CKEditor. I would prefer using our own RichTextEditor for Requirements tab as well once we are able to decouple it from comments completely.

**Comment by maxceem** [comment]
> I would prefer using our own RichTextEditor for Requirements tab as well once we are able to decouple it from comments completely.

Yeah, we can do that for now. Though the week point of our editor is that it doesn't support mobile devices which [other copilots](https://github.com/appirio-tech/connect-app/issues/3213) and I personally would like to be able to use.

**Comment by maxceem** [comment]
6. I've noticed that here we are showing submission ids for on the previous:

     ![image](https://user-images.githubusercontent.com/146016/62687768-034f4500-b9fa-11e9-8985-cf89b9d29037.png)

     Should we add them? If yes, then we should add one more field when we enter designs:

     ![image](https://user-images.githubusercontent.com/146016/62687904-48737700-b9fa-11e9-9832-91bc20c6d1dc.png)

**Comment by vikasrohit** [comment]
> it doesn't support mobile devices which other copilots and I personally would like to be able to use.
Agreed. But we would handle that separately after this big change in the product.

**Comment by vikasrohit** [comment]
6. Yes, we can add a new field for submission id. We might need to create submission preview/download link using that submission id e.g. we can generate a download URL for submission using the Online Review url.

**Comment by maxceem** [comment]
7. When we support this combination of `add-links` and `review` milestones it works like this in `phases`. During `add-links` milestone copilot adds links. After that, when `review` milestone is activated, the data from `add-links` is copied to the `prevMilestoneContent` property of the `review` milestone. And after the `review` milestone uses the list of links which it has inside of `prevMilestoneContent`, not from the previous milestone.

    I think it the new functionality it's better NOT to do like this. We should add links in `design-work` milestone and keep the list of links there without copying it to `review` milestone. The reason for this is that it's very likely that later we would want to edit such list of design links - for example, make some fix in a link or something. And it would be much easier if we keep the data in one place inside `design-work` milestone. I don't see any problem with getting data from `design-work` milestone from the `review` milestone.

    Do you have any reasons to implement it the same way like we did before with `prevMilestoneContent`?

**Comment by maxceem** [comment]
8. For the copilots to be able to enter links during `design-work` the milestone should be active. How we would activate it?

**Comment by vikasrohit** [comment]
7. I would still want to keep the same functionality here because to me it seems a generic approach to let two milestone talk with each other.

8. I think we had a discussion earlier, where we agreed to add buttons to manually mark a milestone completed. So, it would be similar to what we have as of today, marking a milestone complete would activate next milestone. And the first milestone would be activated when we activate the work, using the status dropdown in the top left of the work card shown here https://marvelapp.com/54cf844/screen/58765965

**Comment by maxceem** [comment]
9. @vikasrohit where should we upload files for design previews? 

     ![image](https://user-images.githubusercontent.com/146016/62773704-6a8df780-bad5-11e9-9ac1-4ab6d8d8c7b2.png)

     We have project/product attachments. Though I'm not sure if we should use them.

     As I see, our current file uploading approach consists of 2 steps:
     - Upload file to `https://upload.fs.topcoder.com`, this gives us URL to Topcoder CDN:

         ![image](https://user-images.githubusercontent.com/146016/62773513-e89dce80-bad4-11e9-8467-d1205cd47e5f.png)

     - We don't use the CDN link directly, but instead, we use Amazon S3 link which is signed as the server-side using Project Service `attachments` endpoint.

        ![image](https://user-images.githubusercontent.com/146016/62773786-a7f28500-bad5-11e9-8110-aaedb5ec38ff.png)
     
     So there are two questions:
     1. Would it be ok to directly use link provided by Topcoder CDN service like https://cdn.fs.topcoder.com/pARH0UySg2BJOkl4YgEw ?

     2. If no. Do you think is there any existent functionality we can reuse for files uploading and generating signed Amazon S3 link?

10. When I upload file to the phase in DEV environment. the CDN link work good, but attachments endpoint of Project Service return broken link to file download:

     So if I upload file for a phase on DEV, and after click that file, I couldn't download it. The same functionality works on PROD.

     ![image](https://user-images.githubusercontent.com/146016/62774303-f0f70900-bad6-11e9-9017-94f798e46350.png)

     Do you think something is misconfigured for DEV? Or it's some kind of issue which somehow was introduced on DEV?

**Comment by vikasrohit** [comment]
9. @maxceem I recently used the file attachment feature for supporting file uploads during project creation https://github.com/appirio-tech/connect-app/commit/84fc58943d10a21966b329808a23f88d916bf2f0. So, that might help you in understanding how it should work. Basically, we use FileStack to upload the file from client machine to a temporary AWS S3 storage and then we move the file from temp bucket to final destination bucket when an attachment object is created using project api endpoints. Hope this helps.

10. In dev we have [disabled](https://github.com/topcoder-platform/tc-project-service/blob/dev/src/routes/attachments/create.js#L72) actual file upload to the final bucket, so the s3 does not work in dev. I can enable it it helps us.

**Comment by maxceem** [comment]
@vikasrohit I've just realised, that after customer enter feedback and click `Mark As Completed` the milestone is marked as completed, and there is no way in UI for copilot/PM to see the provided feedback and selected designs.

**Comment by vikasrohit** [comment]
I think the original design intended the individual milestones to be clickable and when we click on on of the milestone, it shows the view of that milestone in the place where we are showing the view for the currently active milestone. Still, that approach would require the copilot/manager to click on the design review milestone to see the feedback which I guess is okay for the time being as we would send an notification email when the milestone is completed and we can embed a deep link to the design review milestone for the copilot/manager users to view the feedback.
Do we have the capability of clicking the individual milestones and see their content?

**Comment by maxceem** [comment]
> Do we have the capability of clicking the individual milestones and see their content?

So far we don't have such capability. And we allow for now only view/interact with currently active milestone. If we follow the milestone link which is now not active, we would be redirected to the work details tab.

Though I think it's not hard to change this behaviour. The only question is:
1. What to show for other types of milestones for which we don't have individual page design at the moment?
2. As I understand we shouldn't be able to click not yet `active` or `completed` milestones, like milestone from the future? What should happen when we click milestones from the future (`planned` or `draft`)?

**Comment by vikasrohit** [comment]
1. What are the milestones which don't yet have designs?
2. I am not sure about it, but it we can just show the planned text for such milestones?

**Comment by maxceem** [comment]
> 1. What are the milestones which don't yet have designs?

- `Phase Specification`
- `Community Work`
- `Final Fixes`

> 2. I am not sure about it, but it we can just show the planned text for such milestones?

Yes, we can show the planned text, though so far we don't have ability to edit it. Should we show it on the milestone edit form? If so, should we also show other texts like active text, completed text and blocked text for consistency?

**Comment by vikasrohit** [comment]
Yep, we can add the fields to the milestone edit form to allow edit of the various texts. And I guess, if we are showing the text for the view of a milestone based on status, we can take care of the 3 milestones' views.
